source/common/router/rds_impl.h

```diff
@@ -193,7 +193,8 @@ struct UpdateOnDemandCallback {
  * the subscription.
  */
 class RdsRouteConfigProviderImpl : public RouteConfigProvider,
-                                   Logger::Loggable<Logger::Id::router> {
+                                   Logger::Loggable<Logger::Id::router>,
+                                   public std::enable_shared_from_this<RdsRouteConfigProviderImpl> {
```

Comment:
can we please not do this. When provider is capable of shared_from_this, we actual open its access to worker threads, which I'd hope we can avoid: providers are supposed to be created/owned by HCM and no one else should have ownership to it, otherwise it'd introduce lots of trouble to do destruction orders when Envoy tears down. 
If you need a cleaner signal to decide if we need a post, one thing you can do is to lock on route_config_updated_cb, which the finest grained object that tells if the source httprequest is alive or not.

Reply:
I don't quite understand this now that I think about it more. This is already a shared_ptr allocated object. Why does it matter if it enabled shared_from_this?

Reply:
I think we want it shared between listeners in main thread, but its ownership should not be shared to worker thread.
If however the ownership of the provider is shared to worker thread, the scenario i was talking about would cause the assertion error when LDS deletes the listener or when envoy quits.

Reply:
I think the change is quite safe: we aren't changing provider interface, just the concrete implementation: to create another shared_ptr one would need to cast a RouteConfigProvider object to RdsRouteConfigProviderImpl before being able to do so. Code review would catch such a cast or a call to `shared_from_this()`. I can add a note re: intended usage of shared_from_this if that would help.

Reply:
I agree that this is the right change modulo any other lifetime issue.

Reply:
it's a tricky case and it's not about life cycle of provider when the post lambda gets called.
 the problem is that The ownership of any "tls instance owner" should not be spilled into other threads than the main thread.
please see my other reply why I am against this, your change is introducing another race which could ends in an assertion failure.